zstd_stream.go

```diff
@@ -205,6 +235,10 @@ func NewReaderDict(r io.Reader, dict []byte) io.ReadCloser {
 
 // Close frees the allocated C objects
 func (r *reader) Close() error {
+	cb := r.compressionBuffer
+	db := r.decompressionBuffer
+	cPool.Put(&cb)
```

Comment:
So...  This probably doesn't matter, but I think this means there is an extra slice allocation/deallocation on each use of this object? E.g.:

1. Call New(): This allocates `&(make([]byte, ...)` which is a pointer to a slice; the slice is a struct that contains bytes. That is:

result ->  slice -> byte array

2. Dereference this into the `reader` struct; I think this effectively does:

reader.slice -> byte array

3. Close: Allocate a new slice:

cb -> slice -> byte array

4. Close: Put that slice into the pool

You could avoid this by making `compressionBuffer` and `decompressionBuffer` *[]byte themselves, which is pretty annoying. Now that I've typed all this: slices are tiny and this probably doesn't matter, and I usually argue for the simpler code so ignore this? Still, if this first change gave you a ~50% improvement, will this give you another ~2-5%?

Reply:
I hear you. I didn't see these causing pain in the pprof dump, but we'll see once this is out in the wild. If so, I'll change `reader.compressionBuffer` and `reader.decompressionBuffer` to be slice pointers in a subsequent PR.